tari-project/sha-p2pool
# feat: add tribes and graceful shutdown

Description
---
We needed tribes to let people mine together in a smaller group called tribe. P2pool must allow this functionality.
Also we should have the chance to list all current tribes, cli of p2pool should support this.
As a plus p2pool should support graceful shutdown.

Motivation and Context
---

How Has This Been Tested?
---

**Testing tribes**

1. Start p2pool node on tribe `a`: `export TARI_NETWORK="esmeralda" && ./target/release/sha_p2pool start -g 18145 --tribe a`
2. Start mining on this p2pool node
3. Start another p2pool node on tribe `b`: `export TARI_NETWORK="esmeralda" && ./target/release/sha_p2pool start -g 18146 --tribe b`
4. Start mining on this other p2pool node
5. It should be visible from the logs that the 2 pools have a different share chain

**Testing list tribes**
Simply run `sha_p2pool list-tribes`
Example output:
```json
["default","a"]
```

What process can a PR reviewer use to test or verify this change?
---
See tests


Breaking Changes
---

- [x] Other - Peer info contains tribe now

BREAKING CHANGE: Seed peers needs to be updated to the latest version as Peer info contains tribe now
